File: haskell/playground/sw.hs
main = do
    print $ hasTargetWindow [1,3,8,10] 4
    print $ hasTargetWindow [10] 10
    print $ hasTargetWindow [0, 10] 10
    print $ hasTargetWindow [0] 10

hasTargetWindow arr target = (tup, rw == target)
    where
        arrLen = length arr
        (left, right, wsum) = (-1, -1, 0)
        tup@(rl, rr, rw) = helper (left, right, wsum) arr

        helper tup [] = tup
        helper tup@(left, right, wsum) arr
            | wsum == target = (left, right, wsum)
            | right + 1 >= arrLen || left + 1 >= arrLen = tup
            | wsum < target = 
                helper (left, right + 1, wsum + ((!!) arr $ right + 1)) arr
            | wsum > target = 
                helper (left + 1, right, wsum - ((!!) arr $ left + 1)) arr

File: python/playground/sw.py
#!/usr/bin/env python3

def has_target_window(arr: [int], target: int) -> (int, int, bool):
    """
        arr: integers are greater than 0
        target: greater than 0
        return: (left side of window, right side of window, target found)
    """
    left, right, wsum = -1, -1, 0
    arr_len = len(arr)
    while wsum != target:
        if wsum < target:
            right += 1
            if right >= arr_len:
                break
            wsum += arr[right]
        elif wsum > target:
            left += 1
            if left >= arr_len:
                break
            wsum -= arr[left]
        print(locals())
    return (left, right, wsum == target)


if __name__ == "__main__":
    # print(has_target_window([1, 3, 8, 10], 18))
    print(has_target_window([10], 10))
    print(has_target_window([0, 10], 10))
